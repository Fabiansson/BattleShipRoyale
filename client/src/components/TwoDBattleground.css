.grid { 
    margin: 1em auto; 
    border-collapse: collapse; 
    /*transform: rotateX(60deg) rotateY(0deg) rotateZ(45deg);*/
}
.grid td {
    cursor: pointer;
    width:60px; height:60px;
    border:1px solid #ccc;
    /*background-image: url(http://img14.deviantart.net/f99b/i/2011/354/d/5/isometric_new_tiles_by_spasquini-d4jnp2i.png);
    background-size: 1600% 100%;*/
    transition: .2s ease-in-out;
   
}
.grid td:hover{
    opacity: .8;
    transform: translateX(-5px) translateY(-5px);
}

.grid td.clicked {
    background-color: red;
}

.tile-0{
    /*background-position: calc(-6px + (-183.5px * 0)) 0px;*/
    background: url('data:image/jpeg;base64,/9j/4AAQSkZJRgABAQAAAQABAAD/2wCEAAkGBxITEhUSEhIVFRUXFRUVFRUVFRUVFRUVFRUWFhUVFRUYHSggGBolHRUVITEhJSkrLi4uFx8zODMtNygtLisBCgoKDg0NFw8PFS0dFR0rLSstKy0rLS0tLS0tLS0tKy0rLS0rLS0rLSstNy0tLSstLS0rLSs3NystLS03LS0rLf/AABEIAMIBAwMBIgACEQEDEQH/xAAbAAACAwEBAQAAAAAAAAAAAAADBAABAgUGB//EADAQAAIBAwMDAgUEAgMBAAAAAAABAgMRIQQxQRJRYXGRBSKBsfATocHRMuEUUvEG/8QAGAEBAQEBAQAAAAAAAAAAAAAAAAECAwT/xAAcEQEBAQEAAwEBAAAAAAAAAAAAARECEiExUUH/2gAMAwEAAhEDEQA/APINhELqfcLBnvleKxtsiyaaKhg0yPTp4N0ykzUUacqMkahSMI3GWAzVulc3SQB1X3IplMpuEcjEYAKEkMXuGK2i0yul2NwDItCXASUQUExhNACUeGYr0cXG6aTLlR8kHKj3C7ha2nBNAZlAVqwHU8GJQCylFEL1GnEnIEignQRBIyAGoWNuAaKTNuGNgjnunnYDW0qfg6n6WMAJUQuuXLRmZaNWzj6HTdJlT07ZGvKuLLQO5DrPTP8ALkGRrzrzEoESJYIkZx6LVI0pIxIkWEO08o0helMOpm3KwVItSwVTfBJIMgVGSnBt2CxV2NaeNmMW9YPp9OM0ad2YUxinCxXG0SqlwYhG67GZ3uMaZAYpuwzGCaCQprJro7bEQDps77Blky6TvvcJRiwMfpNiVamdem7+BXVaTkarnpXwU4NG5qxiVVgAqbmU0bdNsyqYVuWEB6ncIzCgCGqU0M06icZX8fc5E52eH6hoV8P+AY61K1hj/hpq5z/h+rvhrb8yd2l/i2u1zNIQqUkuBb9RRT2OjTj1P0FNZpVn9vAiOZPV5/2QTqaRttpMhpcjiSpgpIMa6SPRuFnEykMOBh0yY1qrBaaYOMWFSYiUeDDQQKihuNTg3HGqjTsrmoJvY31JhU0gzo1CI10iMJZsO0Ihlp0Gw1JW4N0pZGOjGOSBeDuhlRsti4U7FVL28kFNfiMRbuSStj6m0+QiXyEdQH1AZ1GnYK1WprkDOnHhDDdwbgvJUDhFWbEaq7Ds6DatF38bMB+m72aaClVTb4LrQshiisvLF9a77AIPJdONmEpUu4OvhYDTs/D9Pfpts/sd2i8dKOB8LnayWWt7eTuwkzNSJFpSfaxzfiVa8umOy3Z1NRTxjd49DiTXzPt3EKEqXp9SEt6kNMvJxCE6C4kem1bjdApRsMXMVIlSUOLQaCBILAFMRib/AEioSGlC6wVytLQugtJmKqN0AU5SpDsY5FNPMcuGGm+Qmlqu+QDZXW0QdO2SSab/ADsLaetfc29sd2RWZyM3MUp7p8GotXKhh4sB6Lth4QuHjC2fx+hNUnSpu+S4Ry1v6jDje4K6V7gYk3xkLBqas/o+UI6ipnAfR1EsvkWELamg4XOPVnJytbB6LVVk1ZK5x66lZ2X8CAPUkhfUZN6XSyby7d7hanSsO/hrwVr46P8A8tp8SldXvbp553O8lZ2W7X4zzXwyo+rqV9l1eV39T0EINtO90/y5mpplq0UtzlV4LPk61XaxyNSs7jk6Af0ISUokNMvIotpGpIzIO4kUgjigMEGiVKw6JSp2GYstQuE0Hqaew1RnkxKibp07BmpqI5wSETbLTKg1GyYSVRiTmFU2Ew1TqBXU2FI3GqMSIY06C9WTNLCLm0iDM42Zt0+xVOVxiKQFU3ZWKepd7LCCSd8AY0bv0ZFM6hpRTTXlCGpqpqyGnSuc/Wxcfq8CCqUbu/Az+l7GNLEZqYWNxQs10/2L14srVVMC8XUef2uUHSwm/YT1lPZd17DUFm7MwXVPwA18IirvGGrHoNHBRTT7YOPRmoLIfUa9KKsrtrczfaynq042d7HDnDLtfO1wdHUyk7N8jEVnnBZ6SjU9OrLBAinLhYKIPERQaEEApzDRkbda3ZFq1wZpIMisibuTq4NMI2pBEgCZqnLNmEFZgI2DgASEDXQSEAygEXShjyN042QNIKkEFigLy/AXqJ1W2IC0oewzCKBacbpPwZqxnoC06PZfUJhIwqz2wRotK3V2WQGvo9ULfuNRp39eAVfG/crJTRxsrPjYmrqWT2/oLK3+wU4cPlZKOJVrtsVVeae78IJrqLhPP0YGD+/9Fah2jJv1G9PG3qCoJ4t7D9HT5V+E37K5GQ9ZTfQxWinK3g9H+h1w8efVgKel6NktyauEdNo31PHodCVkrWd/GToRji6fHYSqVLPffwTdXMZVF92iGnUfdfuQienzqDGKbwK2GaR0jt0LEtszcpMrAyNXMXKcgipzCUpAmbigGU7hIwBwQxEM0eCCRiATDR2DKpsKsIE0Z628gHlPhF02Bgy41LAdjTrAbqsc/T6ngfhIxWop1u+Cqr4v6sjS8EdO62+pAzTikl6GaivxugcG0rMItr9gpGqrK3bYxSWUH1GVdC9LBpln4zpFJfQ8w6bi36o9tVtJHn6lP5pYzdY+wi7iUOGdig/nh5x7o586bte39juii7Rb3TuvQVI9DpNNddL4uY1OjaljY6GmSsaksnDy9vR4TCUaKtYQr6ZXd0dipJJPvwJVposqdSEVoU8kGut+SjW1jI+UQYaLFqTCOZ0dLBuo1FgVI3ErOCXJySJAjaiajFmIsLErNGp1A8ZikUGjIM01Gog8ZCaYVBDSK6MglILCYRmUS1FBbJmXAC4LlHQ07uhGmxyg7GaQSW+Q1OpYW1Mv4KpywTFOPP1GIrC9hSMrpjNKd4p9mStQo4WuvJqFPBJ73CTkrYKgU30ideCU4y4lh+u6/oZUbsxKj1KUH6oIuNP5fTYPCKXS08WRzdLNqVpfn/p01S6oY9BVjv0+PRe47Frlf0cr4XXvh8I6l3LY8/Ueri7C2rp9reEc2MbvO5068u5x9RWfVh7o3y595o3SQWkn3IaxjXytFtmS0dHYaBrrM0wqSKxWVUNoFcJFhK2i7mVIvqKyIpBadQDY1BBKdjM3EXpoKmGDEZGosApBaYQenI3OV1dcC9Vu2DEagDNKefUfpyOPS+zOjGXYlB6uUn9CaR3VubF9Hyv88i9GeW12z4ZFPxxcNRd0+GCi758AKWotJexFMJcXBQbUrBai5BJZuEOU6S3DOlykCoSu0OyxH1MVuRzNbpox+ayxv6cF6KqndexetkrfcR0M0na+Ua/jP9dem3B3Xo13R2dPqMJpJnDray1rbi0viVubfUzeddOe/F2PiOsjFWvlbvi5x6VZyd7Owu68Zu93L7fQeje1krenBZMjNu0Z34WCAnSf/Z/uQD5aahEqJpSNu9ai7GkwbZpMMrkyQZjqJcGGadjU7AoMFKRWcNcB6FQVhVwajMJYe6zULgYNBVPgrAoegxdBqYZpiQtPDC9VwFTcI33H9KvYQo5Y+pChlSw7A9NTy33K07yF0+3ozIKqvy2Qi4tDMo9hethfUQdHRT6o59DVR2V0hbRu1/I7GN1b89SVpWkqp2d9n7HQrVrr1OdSppPYLVeVYzZ7WXIxXjdM4M6rjM9C3g42qpdUnZf6NRmg1NY2rXsZp0m3l9Xp+YG9N8Ovus/nsdvSfDbZsvv+4vUiyWubp9O8Oyfi+EdXR6Ztpywvz3Go0bbJewxSp87vy7L2Od6dOeApQh2k/Yot6mP/AE/chn236fGEyKRm5EdnTBblSkZRibGpI1c0mZgXcA9KRTjkzEKysIydRTBymDNNU5MahUObCoGjMus3l0VWLVUQUjcJZKx4n4VMmqyFL5CQfFwzhjTsdiKUo2QxTkEo9Odpo6NCG/uc6luNurZYM0jN/mJVhdA07u45FXQGKEXYYUrbdshKKSRLIyuKVRPgk912/NwV2n4Mzbv4KG1Sbt2DafTJt2jhc9zVHEPXA3p2rem+Dna3zylHSeMDCsscL83Mzr4sLSn39v7M+66ep8GnZ8v0vZP18Cyqvqu3hb+oHUahvCx37/6BwpO1kv6XqakYvRi6eXJLwyA46V+SD0nv8fIYs0mCTNtmteqxtywDTyU3ckcDSQW5qINM31FZsFuEWwvBhYyNMWNVJ2QpOQSpIFYzW+ZgikHpSFUw/BYnUHTNwYCnINCWSudhinLIeLFohoM0505CV0EiJuYRVMBnD1GrZ2HXtdHHozydfSzVrvYlRmMmsWuNUlzt6gHXabSVl3ItYiB2EreUW699helqr4ewxTsskDFLT9w89LHlqyBUdRG+br7BJpS5uZutzDqiulEeqjFWE7PjP2IoLnL7cIzjXkPGaefZcsFOLby/oZlWXDySnK+2WVFKSinbJVBuT+aTSRnUNRjl7l0I4vwgn9dH9RdyjnS1cbkJjfk+Qx3CTIQs+PTfrCNEIFWiMhCstwCMhCxm/Q2ZkWQixIh4kIWJ0iC0yENRimkHobP0/khDTjQ1/IWJCAo1Pb3Orp/8SEJWKkuAVZEIVGtIxukyEIpumGpFEMVRrmp7EIRotHceh/iQgqQhrP8AML8RdqWMFEH4v61ooLoWFz92QhDNWP/Z');
    
    background-size: cover;
    
}
.tile-1{
    background-color: rgb(255, 217, 0);
  }
.tile-2{
    background-color: rgb(18, 18, 19);
}
.tile-3{
    background: url('https://cdn4.vectorstock.com/i/1000x1000/81/63/top-view-wood-boat-icon-simple-style-vector-25678163.jpg');
    background-size: cover;
}
.tile-4{
    background-color: rgb(95, 93, 93);
    background-size: cover;
}
.tile-5{
    background-color: rgb(36, 35, 35);
    background-size: cover;
}