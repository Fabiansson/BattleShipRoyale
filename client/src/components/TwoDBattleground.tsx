import React, { useState, useContext } from "react";
import './TwoDBattleground.css';
import { Ship, Fog, Hit, Item } from "../App";

import { useSnackbar } from 'notistack';
import Menu from '@material-ui/core/Menu';
import MenuItem from '@material-ui/core/MenuItem';
import SocketContext from "../services/SocketProvider";
import { coordinateToIndex, calculateDistance } from "../services/helpers";

interface TwoDBattlegroundProps {
    terrain: number[],
    lootMap: number[],
    fog: Fog
    ships: Ship[],
    hits: Hit[],
    inventory: Item[]
}

export interface Move {
    from: number,
    to: string
}

export interface Attack {
    from: number,
    to: number
}

export interface ItemUtilization {
    itemId: number,
    on: number
}

function TwoDBattleground(props: TwoDBattlegroundProps) {
    const { enqueueSnackbar } = useSnackbar();
    const socket = useContext(SocketContext);
    const [selected, setSelected] = useState<string>('');
    const [availableRes, setAvailableRes] = useState<number | null>(null);
    const [menuAnchor, setMenuAnchor] = useState<HTMLTableCellElement | null>(null);
    let mapSize: number = 0;
    let mapWidth: number = 0;

    const createTableRows = () => {
        let mapData: number[] = [...props.terrain];

        mapSize = mapData.length;
        mapWidth = Math.floor(Math.sqrt(mapData.length));

        props.ships.forEach(ship => {
            ship.position.forEach(field => {
                let fieldNumber: number = coordinateToIndex(mapSize, field.x, field.y);
                if (field.health === 1){
                    mapData[fieldNumber] = 3;
                } else if (field.health >= 2) {
                    mapData[fieldNumber] = 7;
                } else {
                    mapData[fieldNumber] = 4;
                }
            })
        })

        props.hits.forEach(hit => {
            let fieldNumber: number = coordinateToIndex(mapSize, hit.x, hit.y);
            mapData[fieldNumber] = 2;
        })

        props.lootMap.forEach(loot => {
            mapData[loot] = 6;
        })

        let table = []
        for (let i = 0; i < mapWidth; i++) {
            let cols = []
            for (let j = 0; j < mapWidth; j++) {
                let tileNumber: number = j + mapWidth * i;
                let tileNumberS: string = tileNumber.toString(10);
                if(isInFog(props.terrain, props.fog, tileNumber, false)) {
                    cols.push(<td id={tileNumberS} key={tileNumberS} className={'tile-5'} onClick={handleFogClick}></td>);
                } else if(isInFog(props.terrain, props.fog, tileNumber, true)) {
                    cols.push(<td id={tileNumberS} key={tileNumberS} className={getBackground(tileNumberS, mapData) + ' soft'} onClick={handleClick}></td>);
                } else {
                    cols.push(<td id={tileNumberS} key={tileNumberS} className={getBackground(tileNumberS, mapData)} onClick={handleClick}></td>);
                }
            }
            table.push(<tr className={'roow-' + i} key={i}>{cols}</tr>)
        }

        return table
    }

    const handleClick = (event: any) => {
        let ship: Ship | null = isShip(event.target.id);
        if (ship) {
            if (selected === event.target.id) {
                setSelected('');
                setAvailableRes(null);
                return;
            }
            setSelected(event.target.id);
            setAvailableRes(ship!.shotsOrMoves);
            setMenuAnchor(event.currentTarget);
            return;
        }
        if (Boolean(selected)) {
            setMenuAnchor(event.currentTarget);
        }
    }

    const handleFogClick = () => {
        enqueueSnackbar('Nothing can be done in the fog...', {
            variant: 'error',
            anchorOrigin: {
              vertical: 'top',
              horizontal: 'center',
          }
          });
    }

    const handleClose = () => {
        setMenuAnchor(null);
    }

    const attack = (anchor: string) => {
        let attack: Attack = { from: parseInt(selected), to: parseInt(anchor) }
        socket?.emit('attack', attack);
        setMenuAnchor(null);
        setSelected('');
        setAvailableRes(null);
    }

    const loot = (anchor: string) => {
        let loot: Attack = { from: parseInt(selected), to: parseInt(anchor) }
        socket?.emit('loot', loot);
        setMenuAnchor(null);
        setSelected('');
        setAvailableRes(null);
    }

    const moveTo = (direction: string) => {
        let move: Move = { from: parseInt(selected), to: direction }
        socket?.emit('moveTo', move);
        setMenuAnchor(null);
        setSelected('');
        setAvailableRes(null);
    }

    const utilizeItem = (itemId: number, anchor: string) => {
        let itemUtilization: ItemUtilization = { itemId: itemId, on: parseInt(anchor) };
        socket?.emit("use", itemUtilization);
        setMenuAnchor(null);
        setSelected('');
        setAvailableRes(null);
      }

    const isShip = (tileNumber: string) => {
        let isShip = null;
        props.ships.forEach(ship => {
            ship.position.forEach(field => {
                let fieldNumber: number = coordinateToIndex(mapSize, field.x, field.y);
                if (fieldNumber === parseInt(tileNumber)) {
                    isShip = ship;
                }
            })
        })
        return isShip;
    }

    const isIsland = (tileNumber: string) =>  {
        return props.terrain[parseInt(tileNumber)] === 1;
    }

    const getBackground = (tileNumber: string, map: number[]) => {
        if (selected === tileNumber) {
            return 'selected';
        } else if (menuAnchor && menuAnchor.id === tileNumber) {
            return 'aimed';
        } 
        return 'tile-' + map[parseInt(tileNumber)];
    }

    function isInFog(map: number[], fog: Fog, field: number, next: boolean) {
        const xCenter = next ? fog.nextXCenter : fog.xCenter;
        const yCenter = next ? fog.nextYCenter : fog.yCenter;
        const radius = next ? fog.nextRadius : fog.radius;

        const fogCenterIndex = coordinateToIndex(map.length, xCenter, yCenter);
        return calculateDistance(map.length, fogCenterIndex, field) > radius;
    }

    return (
        <div>
        <table className="grid">
            <tbody>
                {createTableRows()}
                <Menu
                    id="interactionMenu"
                    anchorEl={menuAnchor}
                    keepMounted
                    open={Boolean(menuAnchor)}
                    onClose={handleClose}
                >
                    {menuAnchor && isShip(menuAnchor.id) && 
                    <React.Fragment>
                    <MenuItem onClick={() => moveTo('up')}>Up</MenuItem>
                    <MenuItem onClick={() => moveTo('down')}>Down</MenuItem>
                    <MenuItem onClick={() => moveTo('left')}>Left</MenuItem>
                    <MenuItem onClick={() => moveTo('right')}>Right</MenuItem>
                    </React.Fragment>}
                    {menuAnchor && !isShip(menuAnchor.id) && !isIsland(menuAnchor.id) &&
                    <MenuItem onClick={() => attack(menuAnchor!.id)}>Attack</MenuItem>}
                    {menuAnchor && !isShip(menuAnchor.id) && isIsland(menuAnchor.id) &&
                    <MenuItem onClick={() => loot(menuAnchor!.id)}>Loot</MenuItem> }
                    {menuAnchor && isShip(menuAnchor.id) && props.inventory.find(item => item.id === 0) &&
                    <MenuItem onClick={() => utilizeItem(0, menuAnchor!.id)}>Protect Ship</MenuItem>}
                    {menuAnchor && isShip(menuAnchor.id) && props.inventory.find(item => item.id === 1) &&
                    <MenuItem onClick={() => utilizeItem(1, menuAnchor!.id)}>Increase Shots</MenuItem>}
                </Menu>
            </tbody>
        </table>
        {availableRes != null && 
        <h3>Available shots or moves with this ship: {availableRes}</h3>}
        </div>
    );
}

export default TwoDBattleground;