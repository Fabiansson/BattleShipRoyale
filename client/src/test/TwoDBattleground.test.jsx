import React from 'react';
import EnzymeToJson from 'enzyme-to-json';
import TwoDBattleground from '../components/TwoDBattleground';
import { SnackbarProvider } from 'notistack';
import {mount} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16'
import { configure } from 'enzyme';

configure({adapter: new Adapter()});
describe('Battleground', () => {
    it('renders correctly', () => {
        const terrainMock = [0,1,0,0,0,0,0,0,0];
        const loopMapMock = [1];
        const fogMock = {
            radius: 2,
            xCenter: 0,
            yCenter: 0,
            nextRadius: 1,
            nextXCenter: 0,
            nextYCenter: 0
        }
        const shipsMock = [{
            shotsOrMoves: 3,
            position: [{
                health: 1,
                x: 2,
                y:2
            }
            ]
        }]
        const hitsMock = [];
        const inventoryMock = [];

        const component = mount(<SnackbarProvider maxSnack={3}>
            <TwoDBattleground terrain={terrainMock} lootMap={loopMapMock} fog={fogMock} ships={shipsMock} hits={hitsMock} inventory={inventoryMock} />
            </SnackbarProvider>)
        expect(EnzymeToJson(component)).toMatchSnapshot();
    });
});