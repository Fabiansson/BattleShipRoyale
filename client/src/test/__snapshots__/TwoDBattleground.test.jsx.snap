// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Battleground renders correctly 1`] = `
<SnackbarProvider
  maxSnack={3}
>
  <TwoDBattleground
    fog={
      Object {
        "nextRadius": 1,
        "nextXCenter": 0,
        "nextYCenter": 0,
        "radius": 2,
        "xCenter": 0,
        "yCenter": 0,
      }
    }
    hits={Array []}
    inventory={Array []}
    lootMap={
      Array [
        1,
      ]
    }
    ships={
      Array [
        Object {
          "position": Array [
            Object {
              "health": 1,
              "x": 2,
              "y": 2,
            },
          ],
          "shotsOrMoves": 3,
        },
      ]
    }
    terrain={
      Array [
        0,
        1,
        0,
        0,
        0,
        0,
        0,
        0,
        0,
      ]
    }
  >
    <div>
      <table
        className="grid"
      >
        <tbody>
          <tr
            className="roow-0"
            key="0"
          >
            <td
              className="tile-0"
              id="0"
              key="0"
              onClick={[Function]}
            />
            <td
              className="tile-6"
              id="1"
              key="1"
              onClick={[Function]}
            />
            <td
              className="tile-0 soft"
              id="2"
              key="2"
              onClick={[Function]}
            />
          </tr>
          <tr
            className="roow-1"
            key="1"
          >
            <td
              className="tile-0"
              id="3"
              key="3"
              onClick={[Function]}
            />
            <td
              className="tile-0 soft"
              id="4"
              key="4"
              onClick={[Function]}
            />
            <td
              className="tile-5"
              id="5"
              key="5"
              onClick={[Function]}
            />
          </tr>
          <tr
            className="roow-2"
            key="2"
          >
            <td
              className="tile-0 soft"
              id="6"
              key="6"
              onClick={[Function]}
            />
            <td
              className="tile-5"
              id="7"
              key="7"
              onClick={[Function]}
            />
            <td
              className="tile-5"
              id="8"
              key="8"
              onClick={[Function]}
            />
          </tr>
          <WithStyles(ForwardRef(Menu))
            anchorEl={null}
            id="interactionMenu"
            keepMounted={true}
            onClose={[Function]}
            open={false}
          >
            <ForwardRef(Menu)
              anchorEl={null}
              classes={
                Object {
                  "list": "MuiMenu-list",
                  "paper": "MuiMenu-paper",
                }
              }
              id="interactionMenu"
              keepMounted={true}
              onClose={[Function]}
              open={false}
            >
              <WithStyles(ForwardRef(Popover))
                PaperProps={
                  Object {
                    "classes": Object {
                      "root": "MuiMenu-paper",
                    },
                  }
                }
                anchorEl={null}
                anchorOrigin={
                  Object {
                    "horizontal": "left",
                    "vertical": "top",
                  }
                }
                getContentAnchorEl={[Function]}
                id="interactionMenu"
                keepMounted={true}
                onClose={[Function]}
                onEntering={[Function]}
                open={false}
                transformOrigin={
                  Object {
                    "horizontal": "left",
                    "vertical": "top",
                  }
                }
                transitionDuration="auto"
              >
                <ForwardRef(Popover)
                  PaperProps={
                    Object {
                      "classes": Object {
                        "root": "MuiMenu-paper",
                      },
                    }
                  }
                  anchorEl={null}
                  anchorOrigin={
                    Object {
                      "horizontal": "left",
                      "vertical": "top",
                    }
                  }
                  classes={
                    Object {
                      "paper": "MuiPopover-paper",
                      "root": "MuiPopover-root",
                    }
                  }
                  getContentAnchorEl={[Function]}
                  id="interactionMenu"
                  keepMounted={true}
                  onClose={[Function]}
                  onEntering={[Function]}
                  open={false}
                  transformOrigin={
                    Object {
                      "horizontal": "left",
                      "vertical": "top",
                    }
                  }
                  transitionDuration="auto"
                >
                  <ForwardRef(Modal)
                    BackdropProps={
                      Object {
                        "invisible": true,
                      }
                    }
                    className="MuiPopover-root"
                    id="interactionMenu"
                    keepMounted={true}
                    onClose={[Function]}
                    open={false}
                  >
                    <ForwardRef(Portal)
                      disablePortal={false}
                    >
                      <Portal
                        containerInfo={
                          <body>
                            <div
                              aria-hidden="true"
                              class="MuiPopover-root"
                              id="interactionMenu"
                              role="presentation"
                              style="position: fixed; z-index: 1300; right: 0px; bottom: 0px; top: 0px; left: 0px; visibility: hidden;"
                            >
                              <div
                                data-test="sentinelStart"
                                tabindex="0"
                              />
                              <div
                                class="MuiPaper-root MuiMenu-paper MuiPopover-paper MuiPaper-elevation8 MuiPaper-rounded"
                                style="opacity: 0; transform: scale(0.75, 0.5625); visibility: hidden;"
                                tabindex="-1"
                              >
                                <ul
                                  class="MuiList-root MuiMenu-list MuiList-padding"
                                  role="menu"
                                  tabindex="0"
                                />
                              </div>
                              <div
                                data-test="sentinelEnd"
                                tabindex="0"
                              />
                            </div>
                          </body>
                        }
                      >
                        <div
                          className="MuiPopover-root"
                          id="interactionMenu"
                          onKeyDown={[Function]}
                          role="presentation"
                          style={
                            Object {
                              "bottom": 0,
                              "left": 0,
                              "position": "fixed",
                              "right": 0,
                              "top": 0,
                              "visibility": "hidden",
                              "zIndex": 1300,
                            }
                          }
                        >
                          <ForwardRef(SimpleBackdrop)
                            invisible={true}
                            onClick={[Function]}
                            open={false}
                          />
                          <TrapFocus
                            disableAutoFocus={false}
                            disableEnforceFocus={false}
                            disableRestoreFocus={false}
                            getDoc={[Function]}
                            isEnabled={[Function]}
                            open={false}
                          >
                            <div
                              data-test="sentinelStart"
                              tabIndex={0}
                            />
                            <ForwardRef(Grow)
                              appear={true}
                              in={false}
                              onEnter={[Function]}
                              onEntering={[Function]}
                              onExited={[Function]}
                              tabIndex="-1"
                              timeout="auto"
                            >
                              <Transition
                                addEndListener={[Function]}
                                appear={true}
                                enter={true}
                                exit={true}
                                in={false}
                                mountOnEnter={false}
                                onEnter={[Function]}
                                onEntered={[Function]}
                                onEntering={[Function]}
                                onExit={[Function]}
                                onExited={[Function]}
                                onExiting={[Function]}
                                tabIndex="-1"
                                timeout={null}
                                unmountOnExit={false}
                              >
                                <WithStyles(ForwardRef(Paper))
                                  className="MuiPopover-paper"
                                  classes={
                                    Object {
                                      "root": "MuiMenu-paper",
                                    }
                                  }
                                  elevation={8}
                                  style={
                                    Object {
                                      "opacity": 0,
                                      "transform": "scale(0.75, 0.5625)",
                                      "visibility": "hidden",
                                    }
                                  }
                                  tabIndex="-1"
                                >
                                  <ForwardRef(Paper)
                                    className="MuiPopover-paper"
                                    classes={
                                      Object {
                                        "elevation0": "MuiPaper-elevation0",
                                        "elevation1": "MuiPaper-elevation1",
                                        "elevation10": "MuiPaper-elevation10",
                                        "elevation11": "MuiPaper-elevation11",
                                        "elevation12": "MuiPaper-elevation12",
                                        "elevation13": "MuiPaper-elevation13",
                                        "elevation14": "MuiPaper-elevation14",
                                        "elevation15": "MuiPaper-elevation15",
                                        "elevation16": "MuiPaper-elevation16",
                                        "elevation17": "MuiPaper-elevation17",
                                        "elevation18": "MuiPaper-elevation18",
                                        "elevation19": "MuiPaper-elevation19",
                                        "elevation2": "MuiPaper-elevation2",
                                        "elevation20": "MuiPaper-elevation20",
                                        "elevation21": "MuiPaper-elevation21",
                                        "elevation22": "MuiPaper-elevation22",
                                        "elevation23": "MuiPaper-elevation23",
                                        "elevation24": "MuiPaper-elevation24",
                                        "elevation3": "MuiPaper-elevation3",
                                        "elevation4": "MuiPaper-elevation4",
                                        "elevation5": "MuiPaper-elevation5",
                                        "elevation6": "MuiPaper-elevation6",
                                        "elevation7": "MuiPaper-elevation7",
                                        "elevation8": "MuiPaper-elevation8",
                                        "elevation9": "MuiPaper-elevation9",
                                        "outlined": "MuiPaper-outlined",
                                        "root": "MuiPaper-root MuiMenu-paper",
                                        "rounded": "MuiPaper-rounded",
                                      }
                                    }
                                    elevation={8}
                                    style={
                                      Object {
                                        "opacity": 0,
                                        "transform": "scale(0.75, 0.5625)",
                                        "visibility": "hidden",
                                      }
                                    }
                                    tabIndex="-1"
                                  >
                                    <div
                                      className="MuiPaper-root MuiMenu-paper MuiPopover-paper MuiPaper-elevation8 MuiPaper-rounded"
                                      style={
                                        Object {
                                          "opacity": 0,
                                          "transform": "scale(0.75, 0.5625)",
                                          "visibility": "hidden",
                                        }
                                      }
                                      tabIndex="-1"
                                    >
                                      <ForwardRef(MenuList)
                                        actions={
                                          Object {
                                            "current": Object {
                                              "adjustStyleForScrollbar": [Function],
                                            },
                                          }
                                        }
                                        autoFocus={true}
                                        autoFocusItem={false}
                                        className="MuiMenu-list"
                                        onKeyDown={[Function]}
                                        variant="selectedMenu"
                                      >
                                        <WithStyles(ForwardRef(List))
                                          className="MuiMenu-list"
                                          onKeyDown={[Function]}
                                          role="menu"
                                          tabIndex={0}
                                        >
                                          <ForwardRef(List)
                                            className="MuiMenu-list"
                                            classes={
                                              Object {
                                                "dense": "MuiList-dense",
                                                "padding": "MuiList-padding",
                                                "root": "MuiList-root",
                                                "subheader": "MuiList-subheader",
                                              }
                                            }
                                            onKeyDown={[Function]}
                                            role="menu"
                                            tabIndex={0}
                                          >
                                            <ul
                                              className="MuiList-root MuiMenu-list MuiList-padding"
                                              onKeyDown={[Function]}
                                              role="menu"
                                              tabIndex={0}
                                            />
                                          </ForwardRef(List)>
                                        </WithStyles(ForwardRef(List))>
                                      </ForwardRef(MenuList)>
                                    </div>
                                  </ForwardRef(Paper)>
                                </WithStyles(ForwardRef(Paper))>
                              </Transition>
                            </ForwardRef(Grow)>
                            <div
                              data-test="sentinelEnd"
                              tabIndex={0}
                            />
                          </TrapFocus>
                        </div>
                      </Portal>
                    </ForwardRef(Portal)>
                  </ForwardRef(Modal)>
                </ForwardRef(Popover)>
              </WithStyles(ForwardRef(Popover))>
            </ForwardRef(Menu)>
          </WithStyles(ForwardRef(Menu))>
        </tbody>
      </table>
    </div>
  </TwoDBattleground>
</SnackbarProvider>
`;
